File: apps/api/encryption.py
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
import secrets
import re
from typing import Optional
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class EncryptionService:
    def __init__(self):
        self.key = None
        self._init_key()
    
    def _init_key(self):
        """Initialize the encryption key from environment variable"""
        base64_key = os.getenv("ENCRYPTION_KEY")
        if not base64_key:
            raise ValueError("ENCRYPTION_KEY environment variable is missing")
        
        # Decode base64 key to bytes
        self.key = base64.b64decode(base64_key)
    
    def _looks_like_plaintext(self, text: str) -> bool:
        """Check if text looks like plaintext (printable ASCII characters)"""
        printable_pattern = re.compile(r'^[\x20-\x7E\s]+$')
        return bool(printable_pattern.match(text)) and len(text) > 0
    
    def _try_extract_plaintext(self, data: str) -> str:
        """Try to extract plaintext from potentially encoded data"""
        if self._looks_like_plaintext(data):
            return data
        
        try:
            decoded = base64.b64decode(data).decode('utf-8')
            if self._looks_like_plaintext(decoded):
                return decoded
        except:
            pass
        
        return data
    
    def encrypt(self, plaintext: str) -> str:
        """Encrypt plaintext using AES-GCM"""
        if not self.key:
            raise ValueError("Encryption key not initialized")
        
        # Generate random IV (12 bytes for GCM)
        iv = secrets.token_bytes(12)
        
        # Create AES-GCM cipher
        aesgcm = AESGCM(self.key)
        
        # Encrypt the data
        ciphertext = aesgcm.encrypt(iv, plaintext.encode('utf-8'), None)
        
        # Combine IV and ciphertext
        combined = iv + ciphertext
        
        # Add version byte (0x01) at the beginning
        versioned_combined = b'\x01' + combined
        
        # Encode to base64
        return base64.b64encode(versioned_combined).decode('utf-8')
    
    def decrypt(self, ciphertext_base64: str) -> str:
        """Decrypt ciphertext using AES-GCM"""
        if not self.key:
            raise ValueError("Encryption key not initialized")
        
        try:
            # Decode base64
            try:
                combined = base64.b64decode(ciphertext_base64)
            except:
                return ciphertext_base64
            
            # Check version byte
            if len(combined) > 0 and combined[0] == 0x01:
                # Remove version byte
                data_without_version = combined[1:]
                
                # Extract IV and ciphertext
                iv = data_without_version[:12]
                ciphertext = data_without_version[12:]
                
                # Create AES-GCM cipher and decrypt
                aesgcm = AESGCM(self.key)
                decrypted = aesgcm.decrypt(iv, ciphertext, None)
                
                return decrypted.decode('utf-8')
            else:
                return self._try_extract_plaintext(ciphertext_base64)
        
        except Exception as e:
            print(f"Decryption failed: {e}")
            return self._try_extract_plaintext(ciphertext_base64)

# Global encryption service instance
encryption_service = EncryptionService()

File: apps/api-ts/src/services/encryption.ts
import crypto from "crypto";
import dotenv from "dotenv";

// Load environment variables
dotenv.config();

export class EncryptionService {
  private key: Buffer;

  constructor() {
    this.key = this.initKey();
  }

  private initKey(): Buffer {
    const base64Key = process.env.ENCRYPTION_KEY;
    if (!base64Key) {
      throw new Error("ENCRYPTION_KEY environment variable is missing");
    }

    // Decode base64 key to bytes
    return Buffer.from(base64Key, "base64");
  }

  private looksLikePlaintext(text: string): boolean {
    // Check if text looks like plaintext (printable ASCII characters)
    const printablePattern = /^[\x20-\x7E\s]+$/;
    return printablePattern.test(text) && text.length > 0;
  }

  private tryExtractPlaintext(data: string): string {
    // Try to extract plaintext from potentially encoded data
    if (this.looksLikePlaintext(data)) {
      return data;
    }

    try {
      const decoded = Buffer.from(data, "base64").toString("utf-8");
      if (this.looksLikePlaintext(decoded)) {
        return decoded;
      }
    } catch {
      // Ignore decode errors
    }

    return data;
  }

  /**
   * Encrypt plaintext using AES-256-CBC
   */
  encrypt(plaintext: string): string {
    if (!this.key) {
      throw new Error("Encryption key not initialized");
    }

    // Generate random IV (16 bytes for CBC)
    const iv = crypto.randomBytes(16);

    // Create AES-256-CBC cipher using createCipheriv
    const cipher = crypto.createCipheriv("aes-256-cbc", this.key, iv);

    // Encrypt the data
    let encrypted = cipher.update(plaintext, "utf8", "base64");
    encrypted += cipher.final("base64");

    // Combine IV and ciphertext
    const combined = Buffer.concat([iv, Buffer.from(encrypted, "base64")]);

    // Add version byte (0x01) at the beginning
    const versionedCombined = Buffer.concat([Buffer.from([0x01]), combined]);

    // Encode to base64
    return versionedCombined.toString("base64");
  }

  /**
   * Decrypt ciphertext using AES-256-CBC
   */
  decrypt(ciphertextBase64: string): string {
    if (!this.key) {
      throw new Error("Encryption key not initialized");
    }

    try {
      // Decode base64
      let combined: Buffer;
      try {
        combined = Buffer.from(ciphertextBase64, "base64");
      } catch {
        return ciphertextBase64;
      }

      // Check version byte
      if (combined.length > 0 && combined[0] === 0x01) {
        // Remove version byte
        const dataWithoutVersion = combined.slice(1);

        // Extract IV and ciphertext
        const iv = dataWithoutVersion.slice(0, 16);
        const ciphertext = dataWithoutVersion.slice(16);

        // Create AES-256-CBC decipher and decrypt
        const decipher = crypto.createDecipheriv("aes-256-cbc", this.key, iv);

        let decrypted = decipher.update(ciphertext, undefined, "utf8");
        decrypted += decipher.final("utf8");

        return decrypted;
      } else {
        return this.tryExtractPlaintext(ciphertextBase64);
      }
    } catch (error) {
      console.error("Decryption failed:", error);
      return this.tryExtractPlaintext(ciphertextBase64);
    }
  }
}

// Global encryption service instance
export const encryptionService = new EncryptionService();
